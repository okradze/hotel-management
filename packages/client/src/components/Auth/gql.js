import gql from 'graphql-tag'

const REFRESH_TOKEN = gql`
    query {
        refreshToken
    }
`

export default REFRESH_TOKEN
