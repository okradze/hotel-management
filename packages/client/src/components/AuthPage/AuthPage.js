import React from 'react'
import './AuthPage.scss'

const LoginForm = React.lazy(() => import('../LoginForm/LoginForm'))
const SignupForm = React.lazy(() => import('../SignupForm/SignupForm'))

const AuthPage = () => {
    const [isLogin, setIsLogin] = React.useState(true)

    return (
        <React.Suspense fallback={<p />}>
            <div className="AuthPage-wrapper">
                <div className="AuthPage-form">
                    {isLogin ? (
                        <LoginForm setIsLogin={setIsLogin} />
                    ) : (
                        <SignupForm setIsLogin={setIsLogin} />
                    )}
                </div>
            </div>
        </React.Suspense>
    )
}

export default AuthPage
