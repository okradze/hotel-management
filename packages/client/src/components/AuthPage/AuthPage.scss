.AuthPage {
    min-height: 100vh;
    display: flex;
    justify-content: center;
    align-items: center;
    background-color: var(--light-grey);
}

.AuthPage-form {
    background-color: var(--white);
    border-radius: var(--border-radius);
    box-shadow: 0 0 3px rgba(var(--dark), 0.15);
    padding: 5rem;

    &__pg {
        margin-top: 2rem;
        font-size: 1.5rem;
        color: rgba(var(--secondary), 1);
        font-weight: 600;
    }
}

.LoginForm {
    display: flex;
    flex-direction: column;

    &__input-group {
        display: flex;
        flex-direction: column;
        margin-bottom: 2rem;
    }
    &__error {
        margin-bottom: 2rem;
    }

    input {
        background-color: var(--light-grey);
        min-width: 35rem;
        flex: 1 1 auto;
        font-size: 1.5rem;
    }

    label {
        margin-bottom: 1rem;
        font-size: 1.4rem;
        color: rgba(var(--primary), 1);
        filter: brightness(108%);
        font-weight: 600;
    }
    button {
        margin-bottom: 2rem;
    }
}
.SignupForm__error {
    margin-top: 1rem;
}
