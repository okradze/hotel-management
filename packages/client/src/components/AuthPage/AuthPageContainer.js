import React, { lazy, Suspense } from 'react'

const LoginForm = lazy(() => import('../LoginForm/LoginForm'))
const SignupForm = lazy(() => import('../SignupForm/SignupForm'))

const AuthPageContainer = () => {
    const [isLoginForm, setIsLoginForm] = React.useState(true)

    return (
        <Suspense fallback={<p />}>
            <div className="AuthPage-form">
                {isLoginForm ? (
                    <LoginForm setIsLogin={setIsLoginForm} />
                ) : (
                    <SignupForm setIsLogin={setIsLoginForm} />
                )}
            </div>
        </Suspense>
    )
}

export default AuthPageContainer
