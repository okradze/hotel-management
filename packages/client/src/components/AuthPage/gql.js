import gql from 'graphql-tag'

export const LOGIN = gql`
    mutation login($email: String!, $password: String!) {
        login(data: { email: $email, password: $password }) {
            createdAt
        }
    }
`

export const SIGNUP = gql`
    mutation signup(
        $email: String!
        $password: String!
        $name: String!
        $phone: String!
    ) {
        createHotel(
            data: {
                name: $name
                phone: $phone
                email: $email
                password: $password
            }
        ) {
            createdAt
        }
    }
`
