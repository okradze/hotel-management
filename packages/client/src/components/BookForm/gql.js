import gql from 'graphql-tag'

export const GET_GUESTS = gql`
    query Guests($query: String!, $limit: Int!) {
        guests(data: { query: $query, limit: $limit }) {
            _id
            name
        }
    }
`

export const GET_ROOMS = gql`
    query Rooms($query: String, $limit: Int!) {
        rooms(data: { query: $query, limit: $limit }) {
            _id
            rate
            roomNumber
            type
        }
    }
`
export const CREATE_BOOKING = gql`
    mutation CreateBooking(
        $guest: ID!
        $room: ID!
        $checkIn: String!
        $checkOut: String!
        $color: String!
    ) {
        createBooking(
            data: {
                guest: $guest
                room: $room
                checkIn: $checkIn
                checkOut: $checkOut
                color: $color
            }
        ) {
            _id
            guest {
                name
                phone
            }
            checkIn
            checkOut
        }
    }
`
