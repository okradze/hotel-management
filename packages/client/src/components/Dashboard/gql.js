import gql from 'graphql-tag'

const GET_DASHBOARD_DATA = gql`
    query($startDate: String!, $limit: Int!, $skip: Int!) {
        rooms(data: { limit: $limit, skip: $skip }) {
            _id
            roomNumber
            bookings(startDate: $startDate) {
                guest {
                    name
                    phone
                }
                color
                checkIn
                checkOut
            }
        }
    }
`

export default GET_DASHBOARD_DATA
