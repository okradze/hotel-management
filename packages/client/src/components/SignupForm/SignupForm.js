import React from 'react'
import { useMutation } from '@apollo/react-hooks'
import { withRouter } from 'react-router-dom'
import { createHotelSchema } from '@hb/common'
import { Context } from '../../context/AuthContext'
import withLoader from '../Loader/Loader'
import { SIGNUP } from '../AuthPage/gql'
import handleValidate from '../../utils/handleValidate'
import Input from '../Input/Input'
import '../AuthPage/AuthPage.scss'

export const SignupForm = ({ history, setIsLogin }) => {
    const { setUser } = React.useContext(Context)
    const [userCredentials, setUserCredentials] = React.useState({
        name: '',
        email: '',
        phone: '',
        password: '',
    })
    const [errors, setErrors] = React.useState({
        name: '',
        email: '',
        phone: '',
        password: '',
    })

    const [signup, { loading }] = useMutation(SIGNUP, {
        variables: {
            ...userCredentials,
        },
        onCompleted: handleCompleted,
        onError: err => {
            if (
                err.message.endsWith('email_taken') ||
                err.message.endsWith('phone_taken')
            ) {
                setErrors({
                    ...errors,
                    email: 'მომხმარებელი უკვე არსებობს',
                })
            }
        },
    })

    const handleChange = (e, path) => {
        handleValidate(
            e.target.value,
            setUserCredentials,
            path,
            setErrors,
            createHotelSchema,
            errors,
        )
    }

    function handleSubmit(e) {
        e.preventDefault()
        try {
            createHotelSchema.validateSync(
                {
                    ...userCredentials,
                },
                {
                    strict: true,
                    abortEarly: false,
                },
            )
            signup()
        } catch (e) {
            if (e.inner) {
                e.inner.forEach(({ path, message }) => {
                    setErrors(prevState => ({
                        ...prevState,
                        [path]: [message],
                    }))
                })
            }
        }
    }

    function handleCompleted({ createHotel }) {
        if (createHotel) {
            setUser({ createAt: createHotel.createdAt })
            history.push('/dashboard')
        }
    }

    const ButtonWithLoader = withLoader(() => 'რეგისტრაცია')

    return (
        <>
            <form onSubmit={e => handleSubmit(e)} className="LoginForm">
                <Input
                    text="სახელი"
                    error={errors.name}
                    value={userCredentials.name}
                    onChange={e => handleChange(e, 'name')}
                    type="text"
                    id="name"
                    placeholder="სასტუმროს სახელი"
                />
                <Input
                    text="ელ-ფოსტა"
                    error={errors.email}
                    value={userCredentials.email}
                    onChange={e => handleChange(e, 'email')}
                    type="text"
                    id="email"
                    placeholder="შეიყვანეთ ელ-ფოსტა"
                />
                <Input
                    text="საკონტაქტო ტელეფონი"
                    error={errors.phone}
                    value={userCredentials.phone}
                    onChange={e => handleChange(e, 'phone')}
                    type="text"
                    id="phone"
                    placeholder="შეიყვანეთ ტელეფონი"
                />
                <Input
                    text="პაროლი"
                    error={errors.password}
                    value={userCredentials.password}
                    onChange={e => handleChange(e, 'password')}
                    type="password"
                    id="password"
                    placeholder="შეიყვანეთ პაროლი"
                />

                <button type="submit" className="button button--secondary">
                    <ButtonWithLoader isLoading={loading} />
                </button>
            </form>
            <button
                type="button"
                onKeyPress={e => {
                    if (e.which === 13) {
                        setIsLogin(true)
                    }
                }}
                onClick={() => setIsLogin(true)}
                className="button--transparent form-link"
            >
                შესვლა &rarr;
            </button>
        </>
    )
}

export default withRouter(SignupForm)
