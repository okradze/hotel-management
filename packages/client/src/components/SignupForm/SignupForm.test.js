import { MockedProvider } from '@apollo/react-testing'
import { SignupForm } from './SignupForm'
import { SIGNUP } from '../AuthPage/gql'

const mocks = [
    {
        request: {
            query: SIGNUP,
            variables: {
                email: 'abc@gmail.com',
                phone: '555555555',
                name: 'Shcellar',
                password: 'password123',
            },
        },
        result: {
            data: {
                createHotel: {
                    createdAt: '2019',
                },
            },
        },
    },
    {
        request: {
            query: SIGNUP,
            variables: {
                email: 'abc@gmail.com',
                phone: '555555555',
                name: 'Shcellar',
                password: 'password123',
            },
        },
        error: new Error('no_user'),
    },
]

describe('SignupForm', () => {
    const setUser = jest.fn()

    jest.spyOn(React, 'useContext').mockImplementation(() => ({
        setUser,
    }))

    it('renders', () => {
        const wrapper = shallow(<SignupForm />)
        expect(wrapper).toMatchSnapshot()
    })

    describe('calls function on', () => {
        let setState, wrapper

        beforeEach(() => {
            setState = jest.fn()
            jest.spyOn(React, 'useState').mockImplementation(init => [
                init,
                setState,
            ])
            wrapper = shallow(<SignupForm />)
        })
        afterEach(() => {
            jest.clearAllMocks()
        })

        it('submit', () => {
            const event = { preventDefault: jest.fn() }
            wrapper.find('form').simulate('submit', event)
            expect(event.preventDefault).toHaveBeenCalled()
        })

        const event = {
            target: {
                value: 'abc',
            },
        }

        const initState = {
            name: '',
            email: '',
            phone: '',
            password: '',
        }

        it('name change', () => {
            wrapper.find('#name').simulate('change', event)
            expect(setState).toHaveBeenCalledWith({
                ...initState,
                name: event.target.value,
            })
        })

        it('phone change', () => {
            wrapper.find('#phone').simulate('change', event)
            expect(setState).toHaveBeenCalledWith({
                ...initState,
                phone: event.target.value,
            })
        })
        it('email change', () => {
            wrapper.find('#email').simulate('change', event)
            expect(setState).toHaveBeenCalledWith({
                ...initState,
                email: event.target.value,
            })
        })

        it('password change', () => {
            wrapper.find('#password').simulate('change', event)
            expect(setState).toHaveBeenCalledWith({
                ...initState,
                password: event.target.value,
            })
        })
    })

    describe('form changes on', () => {
        let setIsLogin, wrapper

        beforeEach(() => {
            setIsLogin = jest.fn()
            wrapper = shallow(<SignupForm setIsLogin={setIsLogin} />)
        })

        it('click', () => {
            wrapper.find('.form-link').simulate('click')
            expect(setIsLogin).toHaveBeenCalledWith(true)
        })
        it('keypress', () => {
            wrapper.find('.form-link').simulate('keypress', {
                which: 13,
            })
            expect(setIsLogin).toHaveBeenCalledWith(true)
        })
    })
})
