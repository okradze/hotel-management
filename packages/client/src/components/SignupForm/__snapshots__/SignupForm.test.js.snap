// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupForm renders 1`] = `
<Fragment>
  <form
    className="LoginForm"
    onSubmit={[Function]}
  >
    <Input
      error=""
      id="name"
      onChange={[Function]}
      placeholder="სასტუმროს სახელი"
      text="სახელი"
      type="text"
      value=""
    />
    <Input
      error=""
      id="email"
      onChange={[Function]}
      placeholder="შეიყვანეთ ელ-ფოსტა"
      text="ელ-ფოსტა"
      type="text"
      value=""
    />
    <Input
      error=""
      id="phone"
      onChange={[Function]}
      placeholder="შეიყვანეთ ტელეფონი"
      text="საკონტაქტო ტელეფონი"
      type="text"
      value=""
    />
    <Input
      error=""
      id="password"
      onChange={[Function]}
      placeholder="შეიყვანეთ პაროლი"
      text="პაროლი"
      type="password"
      value=""
    />
    <button
      className="button button--secondary"
      type="submit"
    >
      <Component
        isLoading={false}
      />
    </button>
  </form>
  <button
    className="button--transparent form-link"
    onClick={[Function]}
    onKeyPress={[Function]}
    type="button"
  >
    შესვლა →
  </button>
</Fragment>
`;
